lib/nethealth/nethealth.go

```diff
@@ -0,0 +1,590 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package nethealth implements a daemonset that when deployed to a kubernetes cluster, will locate and send ICMP echos
+// (pings) to the nethealth pod on every other node in the cluster. This will give an indication into whether the
+// overlay network is functional for pod -> pod communications, and also record packet loss on the network.
+package nethealth
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"reflect"
+	"sort"
+	"time"
+
+	"golang.org/x/net/icmp"
+	"golang.org/x/net/ipv4"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	// heartbeatInterval is the duration between sending heartbeats to each peer. Any heartbeat that takes more
+	// than one interval to respond will also be considered timed out.
+	heartbeatInterval = 1 * time.Second
```

Comment:
Just curious, isn't this too aggressive? If nethealth implements a full mesh, there will constantly be a lot of heartbeats flying around in a large cluster?

Reply:
I don't think so. Say it's running on a 200 node cluster. Each node pings every other node, and ever node pings the local node. So that's 800 packets per second.

An icmp ping request/response is about 98 bytes on the wire, + 50 bytes in vxlan overhead. So 148 bytes * 800 = 118400 bytes/s. 947,200 bps, or still less than 1 megabit on a gigabit or 10gig link. And I suspect gravity will fall over far before hitting 200 nodes.